flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/RowsClauseBoundedOverProcessFunction.scala

```diff
@@ -0,0 +1,331 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.flink.table.runtime.aggregate
+
+import java.lang.{Long => JLong}
+import java.util.{ArrayList, List => JList}
+
+import org.apache.flink.api.common.state._
+import org.apache.flink.api.java.typeutils.RowTypeInfo
+import org.apache.flink.configuration.Configuration
+import org.apache.flink.runtime.state.{FunctionInitializationContext, FunctionSnapshotContext}
+import org.apache.flink.streaming.api.checkpoint.CheckpointedFunction
+import org.apache.flink.streaming.api.functions.ProcessFunction
+import org.apache.flink.table.functions.{Accumulator, AggregateFunction}
+import org.apache.flink.types.Row
+import org.apache.flink.util.{Collector, Preconditions}
+import org.apache.flink.table.TableAPIConfigConstant._
+
+import scala.collection.mutable
+import scala.collection.mutable.PriorityQueue
+
+/**
+  * Process Function for ROWS clause bounded OVER window
+  *
+  * @param aggregates           the list of all [[org.apache.flink.table.functions.AggregateFunction]]
+  *                             used for this aggregation
+  * @param aggFields            the position (in the input Row) of the input value for each aggregate
+  * @param forwardedFieldCount  the count of forwarded fields.
+  * @param aggregationStateType the row type info of aggregation
+  * @param inputType            the row type info for input data
+  * @param processingOffset     the offset of processing
+  * @param isRowTimeType        It is a tag that indicates whether the time type is rowTimeType
+  */
+class RowsClauseBoundedOverProcessFunction(
+    private val aggregates: Array[AggregateFunction[_]],
+    private val aggFields: Array[Int],
+    private val forwardedFieldCount: Int,
+    private val aggregationStateType: RowTypeInfo,
+    private val inputType: RowTypeInfo,
+    private val processingOffset: Int,
+    private val isRowTimeType: Boolean)
+    extends ProcessFunction[Row, Row] with CheckpointedFunction {
+
+  Preconditions.checkNotNull(aggregates)
+  Preconditions.checkNotNull(aggFields)
+  Preconditions.checkArgument(aggregates.length == aggFields.length)
+
+  private var output: Row = _
+  private var accumulators: Row = _
+  private var sameTimeDataBuffer: JList[Row] = _
+
+  // save last expired data
+  private var lastExpiredRow: Row = _
+  // persistent last expired data for recovery
+  private var lastExpiredRowState: ListState[Row] = null
+
+  // for event-time, sorted the data by timestamp
+  private var bufferedRowPriorityQueue: PriorityQueue[(Long, Long)] = _
```

Comment:
A priority queue which keeps the time stamps of the data that are not expired. The first element of the tuple is the origin time stamp (ts) of each data. The second element of the tuple is the derived time stamp (Long.Max - ts), as we want a min-priority queue based on ts. This timeStampPriorityQueue is used for retracting the expired  (out of window scope) data while the over window is sliding.